Core/GDCore/Project/Project.cpp

```diff
@@ -1003,6 +1003,26 @@ bool Project::ValidateName(const gd::String& name) {
   return !(name.find_first_not_of(allowedCharacters) != gd::String::npos);
 }
 
+gd::String Project::GetSafeName(const gd::String& name) {
+  if (name.empty()) return "Unnamed";
+
+  gd::String newName = name;
+
+  if (isdigit(name[0])) newName = "_" + newName;
+
+  gd::String allowedCharacters =
+      "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_";
+  
+  for (size_t i = 0;i < newName.size();++i) {
+    // Replace all unallowed letters by an underscore.
+    if (allowedCharacters.find_first_of(std::u32string(1, newName[i])) == gd::String::npos) {
+      newName.replace(i, 1, '_');
+    }

```

Comment:
What if none of the characters are safe?
Should the renaming be aboorted rather than using `_______`?